File: copilot-silent-chat-vscode/src/store/DatabaseManager.ts
import * as vscode from 'vscode';
import Database from 'better-sqlite3';
import * as path from 'path';
import * as fs from 'fs';

/**
 * DatabaseManager — equivalent of DatabaseManager.kt
 *
 * Raw SQLite via better-sqlite3. Per-workspace DB at globalStorageUri.
 * Schema migration via ALTER TABLE ADD COLUMN (same pattern as Kotlin).
 */
export class DatabaseManager implements vscode.Disposable {
  private db: Database.Database | null = null;

  constructor(private readonly storageUri: vscode.Uri) {}

  getDb(): Database.Database {
    if (!this.db) {
      this.db = this.openDatabase();
    }
    return this.db;
  }

  private openDatabase(): Database.Database {
    const dbDir = this.storageUri.fsPath;
    fs.mkdirSync(dbDir, { recursive: true });

    const dbPath = path.join(dbDir, 'sessions.db');
    const db = new Database(dbPath);

    // WAL mode for concurrent access (same as Kotlin)
    db.pragma('journal_mode = WAL');

    this.initSchema(db);
    return db;
  }

  private initSchema(db: Database.Database): void {
    db.exec(`
      CREATE TABLE IF NOT EXISTS playbook_runs (
        id TEXT PRIMARY KEY,
        start_time INTEGER,
        end_time INTEGER
      );

      CREATE TABLE IF NOT EXISTS chat_sessions (
        session_id TEXT PRIMARY KEY,
        playbook_id TEXT,
        start_time INTEGER,
        end_time INTEGER,
        status TEXT DEFAULT 'ACTIVE',
        title TEXT,
        FOREIGN KEY (playbook_id) REFERENCES playbook_runs(id)
      );

      CREATE TABLE IF NOT EXISTS session_entries (
        id TEXT PRIMARY KEY,
        chat_session_id TEXT NOT NULL,
        entry_type TEXT NOT NULL,
        turn_id TEXT,
        round_id TEXT,
        start_time INTEGER,
        end_time INTEGER,
        status TEXT,
        duration_ms INTEGER,

        -- message fields
        prompt TEXT,
        response TEXT,
        reply_length INTEGER,

        -- tool_call fields
        tool_name TEXT,
        tool_type TEXT,
        input TEXT,
        input_message TEXT,
        output TEXT,
        error TEXT,
        progress_message TEXT,

        -- step fields
        title TEXT,
        description TEXT,

        FOREIGN KEY (chat_session_id) REFERENCES chat_sessions(session_id)
      );
    `);

    // Schema migrations (same pattern as Kotlin: ALTER TABLE ADD COLUMN)
    this.migrateSchema(db);
  }

  private migrateSchema(db: Database.Database): void {
    const migrations: Array<{ table: string; column: string; type: string }> = [
      // Add columns that may be missing from older schemas
      { table: 'chat_sessions', column: 'title', type: 'TEXT' },
      { table: 'session_entries', column: 'round_id', type: 'TEXT' },
      { table: 'session_entries', column: 'progress_message', type: 'TEXT' },
    ];

    for (const { table, column, type } of migrations) {
      try {
        db.exec(`ALTER TABLE ${table} ADD COLUMN ${column} ${type}`);
      } catch {
        // Column already exists — ignore
      }
    }
  }

  dispose(): void {
    this.db?.close();
    this.db = null;
  }
}

File: citi-ai-intellij-plugin/src/main/kotlin/com/citi/assist/store/db/DatabaseManager.kt
package com.citi.assist.store.db

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import java.io.File
import java.sql.Connection
import java.sql.DriverManager

/**
 * Manages SQLite database connections at ~/.citi-ai/projects/{project-slug}/.
 *
 * Uses raw JDBC instead of Exposed ORM to avoid classloader conflicts
 * with IntelliJ's bundled Kotlin/coroutines.
 */
object DatabaseManager {

    private val log = Logger.getInstance(DatabaseManager::class.java)

    private val sessionsConns = java.util.concurrent.ConcurrentHashMap<String, Connection>()

    /**
     * Returns the project data directory: ~/.citi-ai/projects/{project-slug}/
     */
    fun projectDir(project: Project): File {
        val slug = project.name
            .lowercase()
            .replace(Regex("[^a-z0-9]+"), "-")
            .trim('-')
        val dir = File(System.getProperty("user.home"), ".citi-ai/projects/$slug")
        dir.mkdirs()
        return dir
    }

    /**
     * Returns a raw JDBC connection to sessions.db, creating tables if needed.
     */
    fun getSessionsConnection(project: Project): Connection {
        val dbFile = File(projectDir(project), "sessions.db")
        val key = dbFile.absolutePath

        sessionsConns[key]?.let { if (!it.isClosed) return it }
        log.info("Initializing sessions.db at ${dbFile.absolutePath}")

        // Delete stale 0-byte db file from previous failed init
        if (dbFile.exists() && dbFile.length() == 0L) {
            log.warn("Deleting empty sessions.db file from previous failed init")
            dbFile.delete()
        }

        try {
            Class.forName("org.sqlite.JDBC", true, DatabaseManager::class.java.classLoader)
        } catch (e: ClassNotFoundException) {
            throw IllegalStateException("sqlite-jdbc driver not found on plugin classpath", e)
        }

        val conn = DriverManager.getConnection("jdbc:sqlite:${dbFile.absolutePath}")
        conn.autoCommit = true

        conn.createStatement().use { stmt ->
            stmt.executeUpdate("PRAGMA journal_mode=WAL;")
            stmt.executeUpdate("PRAGMA foreign_keys=ON;")

            stmt.executeUpdate("""
                CREATE TABLE IF NOT EXISTS playbook_runs (
                    id TEXT PRIMARY KEY,
                    start_time INTEGER NOT NULL,
                    end_time INTEGER
                )
            """)

            stmt.executeUpdate("""
                CREATE TABLE IF NOT EXISTS chat_sessions (
                    session_id TEXT PRIMARY KEY,
                    playbook_id TEXT REFERENCES playbook_runs(id),
                    start_time INTEGER NOT NULL,
                    end_time INTEGER,
                    status TEXT NOT NULL DEFAULT 'ACTIVE'
                )
            """)

            stmt.executeUpdate("""
                CREATE TABLE IF NOT EXISTS session_entries (
                    id TEXT PRIMARY KEY,
                    chat_session_id TEXT NOT NULL REFERENCES chat_sessions(session_id),
                    entry_type TEXT NOT NULL,
                    turn_id TEXT,
                    start_time INTEGER NOT NULL,
                    end_time INTEGER,
                    status TEXT NOT NULL DEFAULT 'running',
                    duration_ms INTEGER,
                    prompt TEXT,
                    response TEXT,
                    reply_length INTEGER,
                    tool_name TEXT,
                    tool_type TEXT,
                    input TEXT,
                    input_message TEXT,
                    output TEXT,
                    error TEXT,
                    progress_message TEXT,
                    round_id INTEGER,
                    title TEXT,
                    description TEXT
                )
            """)

            // Migrate existing databases — add columns that may be missing
            val migrationColumns = listOf(
                "turn_id TEXT",
                "input_message TEXT",
                "progress_message TEXT",
                "round_id INTEGER",
                "title TEXT",
                "description TEXT",
            )
            for (col in migrationColumns) {
                try {
                    stmt.executeUpdate("ALTER TABLE session_entries ADD COLUMN $col")
                } catch (_: Exception) {
                    // Column already exists — expected for new databases
                }
            }
        }

        log.info("sessions.db initialized successfully")
        sessionsConns[key] = conn
        return conn
    }

    /**
     * Returns the path for vectors.db (raw JDBC + sqlite-vec).
     */
    fun vectorsDbPath(project: Project): File {
        return File(projectDir(project), "vectors.db")
    }
}
